File: sdk/java/src/main/java/com/flagmanagment/sdk/Evaluator.java
package com.flagmanagment.sdk;

/**
 * MurmurHash3 32-bit implementation for deterministic cross-SDK bucketing.
 * This is a pure-Java port matching the canonical x86_32 variant used by
 * spaolacci/murmur3 (Go), mmh3 (Python), and all other FlagManagment SDKs.
 * Seed is fixed to 0 for cross-language consistency.
 */
public class Evaluator {

    private static final int C1 = 0xcc9e2d51;
    private static final int C2 = 0x1b873593;
    private static final int SEED = 0;

    /**
     * Returns a bucket between 0 and 99 (inclusive) for the given targeting key.
     */
    public static int bucketUser(String targetingKey) {
        int hash = murmur3_32(targetingKey.getBytes(java.nio.charset.StandardCharsets.UTF_8), SEED);
        return Math.abs(hash % 100);
    }

    /**
     * MurmurHash3 x86_32 implementation.
     */
    public static int murmur3_32(byte[] data, int seed) {
        int h1 = seed;
        int len = data.length;
        int nblocks = len / 4;

        // body
        for (int i = 0; i < nblocks; i++) {
            int i4 = i * 4;
            int k1 = (data[i4] & 0xff)
                    | ((data[i4 + 1] & 0xff) << 8)
                    | ((data[i4 + 2] & 0xff) << 16)
                    | ((data[i4 + 3] & 0xff) << 24);

            k1 *= C1;
            k1 = Integer.rotateLeft(k1, 15);
            k1 *= C2;

            h1 ^= k1;
            h1 = Integer.rotateLeft(h1, 13);
            h1 = h1 * 5 + 0xe6546b64;
        }

        // tail
        int tail = nblocks * 4;
        int k1 = 0;
        switch (len & 3) {
            case 3:
                k1 ^= (data[tail + 2] & 0xff) << 16;
                // fallthrough
            case 2:
                k1 ^= (data[tail + 1] & 0xff) << 8;
                // fallthrough
            case 1:
                k1 ^= (data[tail] & 0xff);
                k1 *= C1;
                k1 = Integer.rotateLeft(k1, 15);
                k1 *= C2;
                h1 ^= k1;
        }

        // finalization
        h1 ^= len;
        h1 = fmix32(h1);

        return h1;
    }

    private static int fmix32(int h) {
        h ^= h >>> 16;
        h *= 0x85ebca6b;
        h ^= h >>> 13;
        h *= 0xc2b2ae35;
        h ^= h >>> 16;
        return h;
    }
}

File: sdk/dotnet/Evaluator.cs
using System;
using System.Text;

namespace FlagManagment.Sdk
{
    /// <summary>
    /// MurmurHash3 32-bit (x86) implementation for deterministic cross-SDK bucketing.
    /// Matches Go (spaolacci/murmur3), Python (mmh3), and Java implementations.
    /// Seed is fixed to 0 for cross-language consistency.
    /// </summary>
    public static class Evaluator
    {
        private const uint C1 = 0xcc9e2d51;
        private const uint C2 = 0x1b873593;
        private const uint Seed = 0;

        public static int BucketUser(string targetingKey)
        {
            byte[] data = Encoding.UTF8.GetBytes(targetingKey);
            uint hash = MurmurHash3_32(data, Seed);
            return (int)(hash % 100);
        }

        public static uint MurmurHash3_32(byte[] data, uint seed)
        {
            uint h1 = seed;
            int len = data.Length;
            int nblocks = len / 4;

            // body
            for (int i = 0; i < nblocks; i++)
            {
                int i4 = i * 4;
                uint k1 = (uint)(data[i4]
                    | (data[i4 + 1] << 8)
                    | (data[i4 + 2] << 16)
                    | (data[i4 + 3] << 24));

                k1 *= C1;
                k1 = RotateLeft(k1, 15);
                k1 *= C2;

                h1 ^= k1;
                h1 = RotateLeft(h1, 13);
                h1 = h1 * 5 + 0xe6546b64;
            }

            // tail
            int tail = nblocks * 4;
            uint k1Tail = 0;
            switch (len & 3)
            {
                case 3:
                    k1Tail ^= (uint)data[tail + 2] << 16;
                    goto case 2;
                case 2:
                    k1Tail ^= (uint)data[tail + 1] << 8;
                    goto case 1;
                case 1:
                    k1Tail ^= data[tail];
                    k1Tail *= C1;
                    k1Tail = RotateLeft(k1Tail, 15);
                    k1Tail *= C2;
                    h1 ^= k1Tail;
                    break;
            }

            // finalization
            h1 ^= (uint)len;
            h1 = FMix32(h1);

            return h1;
        }

        private static uint RotateLeft(uint value, int count)
        {
            return (value << count) | (value >> (32 - count));
        }

        private static uint FMix32(uint h)
        {
            h ^= h >> 16;
            h *= 0x85ebca6b;
            h ^= h >> 13;
            h *= 0xc2b2ae35;
            h ^= h >> 16;
            return h;
        }
    }
}
